### PR TITLE
PHP: metadata plugin based auth API

### LINKED ISSUE #3906: Design and implement metadata plugin based auth API for PHP
-- start using the C core metadata_plugin functionality
-- implement per-rpc creds C core call_credentials
-- update interop tests to demonstrate the new auth API usage

C# and Node implementations are already done and can be used as a reference. @jtattermusch can provide more details.

This is a blocker for the next release, so please prioritize accordingly.

please add @jboeuf when reviewing the new auth API.

### PR DESCRIPTION
**Pull request by stanley-cheung**
* Fixes #3906 
 * Introduced ```CallCredentials::createFromPlugin($callback)``` function to create a ```CallCredentials``` object
 * The ```CallCredentials``` object can either be ```ChannelCredentials::createComposite``` with a ```ChannelCredentials``` object or directly set on a ```Call``` object via the ```setCredentials``` method.
 * Updated the ```per_rpc_creds``` interop test to demostrate the usage of this new API. 
 * Updated PHP surface API to make sure all 4 request type constructors take a Call options ```$options``` parameter, which will be common in the ```AbstractCall``` class constructor.

### CODE REVIEW

**Comment by stanley-cheung** [comment]
Jenkins failure seems unrelated. Can re-run after review. Thanks

File: tools/jenkins/grpc_interop_php/Dockerfile [hunk 1]
```diff
@@ -100,6 +100,10 @@ RUN /bin/bash -l -c "rvm all do gem install ronn rake"
 RUN curl -sS https://getcomposer.org/installer | php
 RUN mv composer.phar /usr/local/bin/composer
 
+# attempt to force a rebuild of the docker image after this point because
+# Protobuf-PHP codegen has been updated
+RUN echo 1
```
Comment (murgatroid99) [inline tools/jenkins/grpc_interop_php/Dockerfile]: you probably shouldn't leave this here

Comment (stanley-cheung) [inline tools/jenkins/grpc_interop_php/Dockerfile]: yea, but I remove it, somehow Jenkins is smart enough to go back to the cached docker image and we are back to the old state. I will remove this as a separate PR and hopefully that will work.

File: src/php/ext/grpc/call_credentials.c [hunk 2]
```diff
@@ -125,9 +126,105 @@ PHP_METHOD(CallCredentials, createComposite) {
...
+                         void *user_data) {
+  plugin_state *state = (plugin_state *)ptr;
+
+  /* prepare to call the user callback function with info from the
+   * grpc_auth_metadata_context */
+  zval **params[1];
+  zval *arg;
+  zval *retval;
+  MAKE_STD_ZVAL(arg);
+  ZVAL_STRING(arg, context.service_url, 1);
```
Comment (jboeuf) [inline src/php/ext/grpc/call_credentials.c]: Looks like the grpc_auth_metadata_context fields are passed as individual params to the callback. Will it break the API if/when we add more (method name, auth context)?

Comment (stanley-cheung) [inline src/php/ext/grpc/call_credentials.c]: I *just* added the change to pass the entire context to the user callback function. :)

**Comment by murgatroid99** [comment]
All of the PHP tests passed.

**Pull request merged**